test/StorageRent/StorageRent.t.sol

```diff
@@ -0,0 +1,1312 @@
+        uint256 extra = bound(delta, 1, type(uint256).max - totalCost);
+        uint256 value = totalCost + extra;
+
+        vm.deal(address(revertOnReceive), value);
+        vm.prank(address(revertOnReceive));
+        vm.expectRevert(StorageRent.CallFailed.selector);
+        fcStorage.batchRent{value: value}(ids, units);
+    }
+
+    function testFuzzBatchRentRevertsExceedsCapacity(
+        address msgSender,
+        uint256[] calldata _ids,
+        uint16[] calldata _units
+    ) public {
+        // Throw away runs with empty arrays.
+        vm.assume(_ids.length > 0);
+        vm.assume(_units.length > 0);
+
+        // Set a high max capacity to avoid overflow.
+        vm.prank(admin);
+        fcStorage.setMaxUnits(1.6e7);
+        uint256 length = _ids.length <= _units.length ? _ids.length : _units.length;
+        uint256[] memory ids = new uint256[](length);
+        for (uint256 i; i < length; ++i) {
+            ids[i] = _ids[i];
+        }
+        uint256 totalUnits;
+        uint256[] memory units = new uint256[](length);
+        for (uint256 i; i < length; ++i) {
+            units[i] = _units[i];
+            totalUnits += units[i];
+        }
+        vm.assume(totalUnits > 0);
+
+        // Buy all the available units.
+        uint256 maxUnits = fcStorage.maxUnits();
+        uint256 maxUnitsPrice = fcStorage.price(maxUnits);
+        vm.deal(address(this), maxUnitsPrice);
+        fcStorage.rent{value: maxUnitsPrice}(0, maxUnits);
+
+        uint256 totalPrice = fcStorage.price(totalUnits);
+        vm.deal(msgSender, totalPrice);
+        vm.expectRevert(StorageRent.ExceedsCapacity.selector);
+        vm.prank(msgSender);
+        fcStorage.batchRent{value: totalPrice}(ids, units);
+    }
+
+    function testBatchRentCheckedMath() public {
+        uint256[] memory fids = new uint256[](1);
+        uint256[] memory units = new uint256[](1);
+        units[0] = type(uint256).max;
+
+        vm.expectRevert(abi.encodeWithSignature("Panic(uint256)", 0x11));
+        fcStorage.batchRent(fids, units);
+    }
+
+    function testFuzzUnitPriceRefresh(uint48 usdUnitPrice, int256 ethUsdPrice) public {
+        // Ensure Chainlink price is positive
+        ethUsdPrice = bound(ethUsdPrice, 1, type(int256).max);
+
+        priceFeed.setPrice(ethUsdPrice);
+        vm.startPrank(admin);
+        fcStorage.refreshPrice();
+        fcStorage.setPrice(usdUnitPrice);
+        vm.stopPrank();
+        vm.roll(block.number + 1);
+
+        assertEq(fcStorage.unitPrice(), (uint256(usdUnitPrice)).divWadUp(uint256(ethUsdPrice)));
+    }
+
+    function testFuzzUnitPriceCached(uint48 usdUnitPrice, int256 ethUsdPrice) public {
+        // Ensure Chainlink price is positive
+        ethUsdPrice = bound(ethUsdPrice, 1, type(int256).max);
+
+        uint256 cachedPrice = fcStorage.ethUsdPrice();
+
+        priceFeed.setPrice(ethUsdPrice);
+
+        vm.prank(admin);
+        fcStorage.setPrice(usdUnitPrice);
```

Comment:
should we call refresh here to ensure that calling it when the cached price is valid doesn't trigger any unwanted side effects?